docs/installation_and_configuration.md
# Installation and Configuration

Follow these steps to set up and configure the **Senk** project on your device:

## Prerequisites

1. **Hardware Setup:**
   - Ensure you have the Nicla Sense ME properly connected to your device.

2. **Arduino IDE:**
   - Install the Arduino IDE on your development machine if not already installed.

3. **Required Libraries:**
    - Make sure to have the following libraries installed in your Arduino IDE. You can install them using the Library Manager in the Arduino IDE.
        - Nicla_System
        - Arduino_BHY2
        - LittleFileSystem
        - ArduinoBLE
        - mbed
        - File
        - FileSystem

## Installation

1. **Download the Project:**
   - Clone or download the [Senk project from the repository](https://github.com/Magform/senk).

2. **Open Project in Arduino IDE:**
   - Open the Arduino IDE.
   - Load the Senk project by selecting the Senk.ino file.

3. **Configure the Arduino IDE:**
   - Ensure you have the necessary board and port selected in the Arduino IDE.

4. **Configure senk:**
   - Customize Senk according to your preferences by following the [configuration guide](#configuration).

5. **Upload the Code:**
   - Upload the Senk code to your Nicla Sense ME.

## Configuration

The Senk project provides a range of configuration options, allowing you to customize its functionality based on your specific needs.\
To make adjustments, locate the `Configuration.h` file in the project's root directory. \
Options:

### Debugging Settings

- **DEBUG_STATUS:**
  - Set to `1` to activate debugging, `0` to deactivate.

### Data Configuration

Before configuring the data part, it's beneficial to understand how datasets function in Senk. \
You can learn more about datasets [here](https://senk.nicolasferraresso.dev/#/implementation_analysis?id=dataset-management).

- **DATA_PER_SET:**
  - Total number of data in a dataset.
  - For better performace is suggested to keep this lower than `MAX_DATASET_DIMENSION`

- **DATA_DISTANCE:**
  - Time interval in milliseconds between two consecutive sets of data.
  - The first data set is captured at system startup.
  - Under normal conditions (with `DEBUG_STATUS` disabled), the error in data capture timing is typically less than 0.5 millisecond. However, when `DEBUG_STATUS` is enabled, the timing error increases to approximately 15-16 milliseconds.

- **DISTANCE_BETWEEN_SET:**
  - Milliseconds to wait between two different datasets. Measured from the first data of the current dataset to the last data of the previous dataset.

- **MAX_DATASET_DIMENSION:**
  - If a dataset surpasses a certain size, it cannot fit entirely into RAM. To accommodate larger datasets, we automatically break them into manageable parts based on the dimension specified here.
  - After processing each part of the complete dataset, the data is dynamically managed – either stored in local storage or transmitted using Bluetooth Low Energy (BLE), contingent on your configuration settings.
  - It's crucial to acknowledge that this segmentation process introduces a minor slowdown, especially when saving data to local storage. Take this into consideration while optimizing for swift data processing.
  - Note: When using all functionalities simultaneously, it is racommended to set this value equal or less than 150.
  - If `SEND_DATASET_THREAD` is enable this value need to be reduced to 10.
  - Adjusting this value becomes crucial when aiming for optimal performance.

### Data Saving Configuration

- **DATA_SAVER:**
  - `1` enables the functionality to save data to local storage; `0` disables it.
  - Manages the opening and closing of the file before and after writing every batch of `DATA_PER_ITERATION` data entries to the file, respectively.

- **DATA_SAVER_KEEP_OPEN:**
  - `1` enables continuous data saving to local storage without closing the file after each write operation.
  - Maintains the file open to write all available data from the dataSet.
  - This mode is incompatible if `DATA_SAVER` is enabled.
  - Faster than `DATA_SAVER`, but if the device loses power while the file is open, data may be lost or corrupted.

- **DATA_PER_ITERATION:**
  - Divides the complete dataSet into groups, with a maximum of `DATA_PER_ITERATION` data entries in each group for writing to the file.
  - Higher values of `DATA_PER_ITERATION` speed up the saving process but consume more RAM.
  - To prevent exceeding memory constraints, ensure `DATA_PER_ITERATION` doesn't exceed 15 with a `MAX_DATASET_DIMENSION` of 150.

- **USER_ROOT:**
  - Specifies the root directory where the file will be saved. It is essential for this path to begin with "fs" to ensure proper file system navigation and storage functionality.
  - The value should be enclosed within double quotation marks ("").

- **DELETE_FILE:**
  - Set to `1` to delete all files and only keep new data.

- **SAVE_FILE_NAME:**
  - Name of the file containing all saved data.
  - The value should be enclosed within double quotation marks ("").

### Data Transmission configuration

- **SEND_DATASET:**
  - Set to `1` to send every dataset to Bluetooth after acquisition.

- **SEND_DATASET_THREAD:**
  - Set to 1, initiates a dedicated thread for transmitting every dataset to Bluetooth after acquisition.
  - Not functional if SEND_DATASET is enabled.
  - Although slightly more energy-consuming than SEND_DATASET, it significantly reduces the delay between two distinct dataset parts when the dataset size exceeds `MAX_DATASET_DIMENSION`.

- **DATA_SENDER:**
  - This option enables reading data from the file and sending it through Bluetooth. It functions only when `DATA_SAVER_STATUS` is enabled. The program scans and sends all available data once during its execution.

- **DATA_TO_SCAN:**
  - Specifies the total lines to read and send in each file scan. If this value exceeds the remaining data to scan, the sender transmits all remaining data.
  - Be cautious, as setting it larger than necessary may result in slower performance. It is recommended to keep this value optimized for efficient processing.

- **SCAN_TIME:**
  - Sets the interval between two consecutive file scans.
  - The specified time is taken at the start of the previous scan.
  - If new data is added during this interval, it will be sent.

- **SERVICE_UUID:**
  - UUID for the main service used in the BLE communication

- **ACCELEROMETER_UUID:**
  - UUID for the BLE characteristic representing accelerometer data

- **GYROSCOPE_UUID:**
  - UUID for the BLE characteristic representing gyroscope
  

power consumption testing/power_consumption_configuration.md
# Power consumption

## Test
Details
- The current implementation uses delay() as a sleep function between two different dataSet which does not cause a true deep sleep in the microcontroller. Because of this we added a theoretical part to the tests where we hypothesize how much the consumption would be if:
    - the sleep state between one dataset and another gave a consumption of approximately 0.2mA and 0.66mW
    - the dataset taking give a consumption of 9mA and 30mW.
    - the dataset saving give a consumption of 10mA and 35mW for iteration/3 seconds
    - the dataset saving keeping the file open give a consumption of 10mA and 35mW for 3 seconds for evry saving process
- With some tests it was noticed that the device has an initial transient, where consumption increases as time passes, quite long so the data was taken for 10 minutes and after this initial transient.
- Measurements conducted using an Otii Arc Pro for accuracy

### Test 1
One dataSet of one data evry 10 seconds
#### Theoretical
`working_time = dataSet * dataDistance = 1*1 = 1ms`  
`theoretical_consumption = (sleeping_time * sleeping_current + working_time * working_current)/(sleeping_time + working_time) = (10000ms * 0.2mA + 1ms * 9mA)/10001ms = 0.2009mA`  
`theoretical_power = (sleeping_time * sleeping_power + working_time * working_power)/(sleeping_time + working_time) = (10000ms * 0.66mW + 1ms * 27.8mW)/10001ms = 0.6630mW`  
#### Pratical
##### Current
Max: 13.5mA  
AVG: 8.42mA  
Min: 5.17mA  
##### Power
Max: 44.3mW  
Avg: 27.8mW  
Min: 17.4mW  


### Test 2 | REDO CHANGED 150 | 1 | 100
One dataSet of 50 data with a distance of 1ms between the data evry 10sec
#### Theoretical
`working_time = dataSet * dataDistance = 50*1 = 50ms`  
`theoretical_consumption = (sleeping_time * sleeping_current + working_time * working_current)/(sleeping_time + working_time) = (10000ms * 0.2mA + 50ms * 9mA)/10050ms = 0.2438mA`  
`theoretical_power = (sleeping_time * sleeping_power + working_time * working_power)/(sleeping_time + working_time) = (10000ms * 0.66mW + 50ms * 30mW)/10050ms = 0.8060mW`  
#### Pratical
Transient of about 10 minutes
##### Current
Max: 13.6mA  
AVG: 8.45mA  
Min: 5.53mA  
##### Power
Max: 44.8mW  
Avg: 27.9mW  
Min: 18.7mW  

### Test 3
One dataSet of 10000 data with 1ms of distance between the data evry 10 seconds
#### Theoretical
`working_time = dataSet * dataDistance = 10000*1 = 10000ms`  
`theoretical_consumption = (sleeping_time * sleeping_current + working_time * working_current)/(sleeping_time + working_time) = (10000ms * 0.2mA + 10000ms * 9mA)/20000ms = 4.6mA`  
`theoretical_power = (sleeping_time * sleeping_power + working_time * working_power)/(sleeping_time + working_time) = (10000ms * 0.66mW + 10000ms * 30mW)/20000ms = 15.33mW`  
#### Pratical
##### Current
Max: 13.6mA  
AVG: 8.54mA  
Min: 5.23mA  
##### Power
Max: 45.1mW  
Avg: 28.2mW  
Min: 17.6mW  

### Test 4
One dataSet of 10000 data with a distance of 1ms between the data evry 100 seconds
#### Theoretical
`working_time = dataSet * dataDistance = 10000*1 = 10000ms`  
`theoretical_consumption = (sleeping_time * sleeping_current + working_time * working_current)/(sleeping_time + working_time) = (100000ms * 0.2mA + 10000ms * 9mA)/110000ms = 1mA`  
`theoretical_power = (sleeping_time * sleeping_power + working_time * working_power)/(sleeping_time + working_time) = (100000ms * 0.66mW + 10000ms * 30mW)/110000ms = 3.33mW`  
#### Pratical
##### Current
Max: 13.5mA  
AVG: 8.33mA  
Min: 5.39mA  
##### Power
Max: 44.5mW  
Avg: 27.5mW  
Min: 18.3mW  

### Test 5
One dataSet of 10000 data with a distance of 10ms between the data evry 10 seconds
#### Theoretical
`working_time = dataSet * dataDistance = 10000*10 = 100000ms`  
`theoretical_consumption = (sleeping_time * sleeping_current + working_time * working_current)/(sleeping_time + working_time) = (10000ms * 0.2mA + 100000ms * 9mA)/110000ms = 8.2mA`  
`theoretical_power = (sleeping_time * sleeping_power + working_time * working_power)/(sleeping_time + working_time) = (10000ms * 0.66mW + 100000ms * 30mW)/110000ms = 27.33mW`  
#### Pratical
##### Current
Max: 13.6mA  
AVG: 8.51mA  
Min: 5.36mA  
##### Power
Max: 44.8mW  
Avg: 28.1mW  
Min: 18.1mW  

### Test 6
One dataSet of 10000 data with a distance of 1ms between the data evry 10 seconds
#### Theoretical
`working_time = dataSet * dataDistance = 10000*1 = 10000ms`  
`theoretical_consumption = (sleeping_time * sleeping_current + working_time * working_current)/(sleeping_time + working_time) = (10000ms * 0.2mA + 10000ms * 9mA)/20000ms = 4.6mA`  
`theoretical_power = (sleeping_time * sleeping_power + working_time * working_power)/(sleeping_time + working_time) = (10000ms * 0.66mW + 10000ms * 30mW)/20000ms = 15.33mW`  
#### Pratical
##### Current
Max: 14.1mA  
AVG: 8.51mA  
Min: 5.19mA  
##### Power
Max: 46.6mW  
Avg: 28.1mW  
Min: 17.5mW  

### Test 7
One datSet of 10000 data with a distance of 10ms between the data evry 100sec
#### Theoretical
`working_time = dataSet * dataDistance = 10000*10 = 100000ms`  
`theoretical_consumption = (sleeping_time * sleeping_current + working_time * working_current)/(sleeping_time + working_time) = (100000ms * 0.2mA + 100000ms * 9mA)/200000ms = 4.6mA`  
`theoretical_power = (sleeping_time * sleeping_power + working_time * working_power)/(sleeping_time + working_time) = (100000ms * 0.66mW + 100000ms * 30mW)/200000ms = 15.33mW`  
#### Pratical
##### Current
Max: 13.7mA  
AVG: 8.59mA  
Min: 5.59mA  
##### Power
Max: 45.4mW  
Avg: 28.4mW  
Min: 19.1mW  

### Test 8
One dataSet of 150 data with a distance of 1ms between the data evry 100sec
dataSaver enable and 1 dataPerIteration
#### Theoretical
`working_time = dataSet * dataDistance = 150*1 = 150ms`  
`saving_time = (dataSet/datePerIteration)*1000/3 = 150*333ms = 50000ms`   
`theoretical_consumption = ( (sleeping_time-saving_time) * sleeping_current + working_time * working_current + saving_current * saving_time)/(sleeping_time + working_time) = (50000ms * 0.2mA + 150ms * 9mA + 50000ms * 10mA)/100150ms = 5.10mA`  
`theoretical_power = ( (sleeping_time-saving_time) * sleeping_power + working_time * working_power)/(sleeping_time + working_time) = (50000ms * 0.66mW + 150ms * 30mW + 50000ms * 35mW)/100150ms = 17.848mW`  
#### Pratical
##### Current
Max: 18.3mA  
AVG: 7.36mA  
Min: 4.22mA  
##### Power
Max: 60.6mW  
Avg: 24.3mW  
Min: 14.2mW  

### Test 9
One dataSet of 150 data with a distance of 1ms between the data evry 100sec
dataSaverKeepOpen enable and 1 dataPerIteration
#### Theoretical
`working_time = dataSet * dataDistance = 150*1 = 50ms`  
`saving_time = 3000ms`   
`theoretical_consumption = ( (sleeping_time-saving_time) * sleeping_current + working_time * working_current + saving_current * saving_time)/(sleeping_time + working_time) = (97000ms * 0.2mA + 150ms * 9mA + 3000ms * 10mA)/100150ms = 0.51mA`  
`theoretical_power = ( (sleeping_time-saving_time) * sleeping_power + working_time * working_power)/(sleeping_time + working_time) = (97000ms * 0.66mW + 150ms * 30mW + 3000ms * 35mW)/100150ms = 1.733mW`   
#### Pratical
##### Current
Max: 14.9mA  
AVG: 5.99mA  
Min: 4.17mA  
##### Power
Max: 49.3mW  
Avg: 19.8mW  
Min: 13.9mW  

### Test 10
One dataSet of 150 data with a distance of 1ms between the data evry 100sec
dataSaver enable and 15 dataPerIteration
#### Theoretical
`working_time = dataSet * dataDistance = 150*1 = 150ms`  
`saving_time = (dataSet/datePerIteration)*1000/3 = ms = 5000ms`   
`theoretical_consumption = ( (sleeping_time-saving_time) * sleeping_current + working_time * working_current + saving_current * saving_time)/(sleeping_time + working_time) = (95000ms * 0.2mA + 150ms * 9mA + 5000ms * 10mA)/100150ms = 0.702mA`  
`theoretical_power = ( (sleeping_time-saving_time) * sleeping_power + working_time * working_power)/(sleeping_time + working_time) = (95000ms * 0.66mW + 150ms * 30mW + 5000ms * 35mW)/100150ms = 2.418mW`   
#### Pratical
##### Current
Max: 17.1mA  
AVG: 6.04mA  
Min: 17.1mA  
##### Power
Max: 56.5mW  
Avg: 20mW  
Min: 13.9mW  

### Test 11
One dataSet of 150 data with a distance of 1ms between the data evry 100sec
dataSaverKeepOpen enable and 15 dataPerIteration
#### Theoretical
`working_time = dataSet * dataDistance = 150*1 = 50ms`  
`saving_time = 3000ms`   
`theoretical_consumption = ( (sleeping_time-saving_time) * sleeping_current + working_time * working_current + saving_current * saving_time)/(sleeping_time + working_time) = (97000ms * 0.2mA + 150ms * 9mA + 3000ms * 10mA)/100150ms = 0.51mA`  
`theoretical_power = ( (sleeping_time-saving_time) * sleeping_power + working_time * working_power)/(sleeping_time + working_time) = (97000ms * 0.66mW + 150ms * 30mW + 3000ms * 35mW)/100150ms = 1.733mW` 
#### Pratical
##### Current
Max: 15.4mA  
AVG: 5.66mA  
Min: 3.86mA  
##### Power
Max: 51.1mW  
Avg: 18.7mW  
Min: 13.0mW  

### Test 12
One dataSet of 10000 data with 1ms of distance between the data evry 10 seconds
send DataSet enable
For all the test a device was connected ad receiving data
#### Theoretical

#### Pratical
##### Current
Max: 13.9mA  
AVG: 8.50mA  
Min: 5.18mA  
##### Power
Max: 45.8mW  
Avg: 28.0mW  
Min: 17.7mW  

### Test 13
One dataSet of 10000 data with 1ms of distance between the data evry 10 seconds
send DataSetThread enable
MAX_DATASET_DIMENSION reduced to 10 to give space for the Thread
For all the test a device was connected ad receiving data
#### Theoretical

#### Pratical
##### Current
Max: mA  
AVG: mA  
Min: mA  
##### Power
Max: mW  
Avg: mW  
Min: mW  

### Test 14

One dataSet of 150 data with a distance of 1ms between the data evry 10sec
dataSaver enable and 15 dataPerIteration
enable dataSender with 150 dataToScan and scanTime of 10sec

### Test 15

One dataSet of 150 data with a distance of 1ms between the data evry 10sec
dataSaver enable and 15 dataPerIteration
enable dataSender with 15000 dataToScan and scanTime of 10sec

### Test 16

One dataSet of 150 data with a distance of 1ms between the data evry 10sec
dataSaver enable and 15 dataPerIteration
enable dataSender with 15000 dataToScan and scanTime of 1000sec

### Test 17

One dataSet of 150 data with a distance of 1ms between the data evry 10sec
dataSaver enable and 15 dataPerIteration
enable dataSender with 150 dataToScan and scanTime of 1000sec

### Test 18
MAX energy

## Conclusion

### Bigger vs smaller dataSet
T1 - T2 - T3

### Bigger vs smaller data_distance

### Bigger vs smaller distance_between_set

### Disable vs enable dataSaver

### Disable vs enable dataSaverKeepOpen

### Disable vs enable sendDataSet

### Disable vs enable sendDataSet Thread

### Disable vs enable dataSender

### Bigger vs smaller dataToScan

### Bigger vs smaller scanTime

### Bigger amout of power consumption


Senk/Configuration.h
#ifndef Configuration_H_
#define Configuration_H_

// FOR INFO ABOUT THE CONFIGURATION VISIT:
// https://senk.nicolasferraresso.dev/#/installation_and_configuration


// Debugging settings
#define DEBUG_STATUS 0


// Data configuration
#define DATA_PER_SET 10000
#define DATA_DISTANCE 1
#define DISTANCE_BETWEEN_SET 0 
#define MAX_DATASET_DIMENSION 10

// Data saving configuration
#define DATA_SAVER 0
#define DATA_SAVER_KEEP_OPEN 0 
#define DATA_PER_ITERATION 15
#define USER_ROOT "fs"
#define DELETE_FILE 1
#define SAVE_FILE_NAME "out.csv"

// Data sending configuration
#define SEND_DATASET 0
#define SEND_DATASET_THREAD 1

#define DATA_SENDER 0
#define DATA_TO_SCAN 0
#define SCAN_TIME 0

// Bluetooth configuration
#define SERVICE_UUID "11111111-0001-1111-1111-111111111111" // UUID for the main service used in the BLE communication
#define ACCELEROMETER_UUID "11111111-0001-1111-1111-111111111111" // UUID for the BLE characteristic representing accelerometer data
#define GYROSCOPE_UUID "11111111-0002-1111-1111-111111111111" // UUID for the BLE characteristic representing gyroscope data


// Debug printing function

#include <rtos.h>
#include <Arduino.h>
inline void debugPrint(const char* toPrint){
  #if DEBUG_STATUS
    static rtos::Semaphore serialSem(1);
    serialSem.acquire();
    Serial.println(toPrint);
    serialSem.release();
  #endif
}

#endif //Configuration_H_